File: 2023/day16/day16.py
#!/usr/bin/env python3
# https://adventofcode.com/2023/day/16

from pathlib import Path
from copy import deepcopy
from collections import defaultdict, deque, namedtuple
import sys, re, math, itertools, time
from functools import reduce
import re

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
lines = data.splitlines()

grid = [[c for c in line] for line in lines]

sx = len(grid[0])
sy = len(grid)

beams = None


def reset_beams():
    global beams
    beams = [[0 for _ in range(sx)] for _ in range(sy)]


def show(show_beams=True):
    n = math.ceil(math.log10(sx))
    for i in range(n):
        print("     " + "".join(f"{x:>{n}}"[i] for x in range(sx)))
    print()

    for y in range(sy):
        s = f"{y:>3}  "
        for x in range(sx):
            e = grid[y][x]
            if show_beams:
                if e == ".":
                    e = beams[y][x]
                    e = ".^v2<222>2222222"[e]
                    e = "\033[1;33m" + e + "\033[0m"
                else:
                    if beams[y][x] != 0:
                        e = "\033[1;33m" + e + "\033[0m"
                    else:
                        e = "\033[1;34m" + e + "\033[0m"
            s += e
        print(s)
    print()


sys.setrecursionlimit(100_000)


def beam(x, y, d):
    if x < 0 or y < 0 or x >= sx or y >= sy:
        return

    id = 1 << "^v<>".index(d)
    if (beams[y][x] & id) == id:
        return

    beams[y][x] = beams[y][x] + id

    if grid[y][x] == ".":
        if d == ">":
            beam(x + 1, y, d)
        elif d == "<":
            beam(x - 1, y, d)
        elif d == "v":
            beam(x, y + 1, d)
        elif d == "^":
            beam(x, y - 1, d)
        else:
            raise ValueError

    elif grid[y][x] == "\\":
        if d == ">":
            beam(x, y + 1, "v")
        elif d == "<":
            beam(x, y - 1, "^")
        elif d == "v":
            beam(x + 1, y, ">")
        elif d == "^":
            beam(x - 1, y, "<")
        else:
            raise ValueError

    elif grid[y][x] == "/":
        if d == ">":
            beam(x, y - 1, "^")
        elif d == "<":
            beam(x, y + 1, "v")
        elif d == "v":
            beam(x - 1, y, "<")
        elif d == "^":
            beam(x + 1, y, ">")
        else:
            raise ValueError

    elif grid[y][x] == "-":
        if d == ">":
            beam(x + 1, y, d)
        elif d == "<":
            beam(x - 1, y, d)
        elif d == "v" or d == "^":
            beam(x - 1, y, "<")
            beam(x + 1, y, ">")
        else:
            raise ValueError

    elif grid[y][x] == "|":
        if d == ">" or d == "<":
            beam(x, y - 1, "^")
            beam(x, y + 1, "v")
        elif d == "v":
            beam(x, y + 1, d)
        elif d == "^":
            beam(x, y - 1, d)
        else:
            raise ValueError
    else:
        raise ValueError


def energized():
    return sum(1 for x in range(sx) for y in range(sy) if beams[y][x] != 0)


# part 1
reset_beams()
beam(0, 0, ">")

if verbose:
    show()
    print(energized())
    exit()

print(energized())

# part 2
m = 0
for x in range(sx):
    reset_beams()
    beam(x, 0, "v")
    m = max(m, energized())
    reset_beams()
    beam(x, sy - 1, "^")
    m = max(m, energized())
for y in range(sy):
    reset_beams()
    beam(0, y, ">")
    m = max(m, energized())
    reset_beams()
    beam(sx - 1, y, "<")
    m = max(m, energized())
print(m)

File: 2023/day16/day16.rs
//! [Day 16: The Floor Will Be Lava](https://adventofcode.com/2023/day/16)

struct Puzzle {
    mirrors: Vec<Vec<char>>,
    beams: Vec<Vec<u8>>,
    sx: usize,
    sy: usize,
}

impl Puzzle {
    fn new() -> Puzzle {
        Puzzle {
            mirrors: vec![],
            beams: vec![],
            sx: 0,
            sy: 0,
        }
    }

    /// Get the puzzle input.
    fn configure(&mut self, path: &str) {
        let data = std::fs::read_to_string(path).unwrap();

        self.mirrors.push(vec![]);
        for line in data.lines() {
            let mut row = vec![];
            row.push(' ');
            for c in line.chars() {
                row.push(c);
            }
            self.mirrors.push(row);
        }
        self.sx = self.mirrors[1].len() - 1;
        self.sy = self.mirrors.len() - 1;
    }

    /// Set beams state to initial value.
    fn reset_beams(&mut self) {
        self.beams.clear();
        for _y in 0..=self.sy {
            let row = vec![0; self.sx + 1];
            self.beams.push(row);
        }
    }

    /// Compute recursively the path of beam.
    fn beam(&mut self, x: usize, y: usize, d: char) {
        // point inside the grid ?
        if x == 0 || y == 0 || x > self.sx || y > self.sy {
            return;
        }

        // if point has already traversed by the beam, stop
        let id = 1 << u8::try_from("^v<>".find(d).unwrap()).unwrap();
        if (self.beams[y][x] & id) == id {
            return;
        }
        self.beams[y][x] |= id;

        match self.mirrors[y].get(x) {
            Some('.') => match d {
                '>' => self.beam(x + 1, y, d), // traverse
                '<' => self.beam(x - 1, y, d), // traverse
                'v' => self.beam(x, y + 1, d), // traverse
                '^' => self.beam(x, y - 1, d), // traverse
                _ => panic!(),
            },

            Some('\\') => match d {
                '>' => self.beam(x, y + 1, 'v'), // reflect
                '<' => self.beam(x, y - 1, '^'), // reflect
                'v' => self.beam(x + 1, y, '>'), // reflect
                '^' => self.beam(x - 1, y, '<'), // reflect
                _ => panic!(),
            },

            Some('/') => match d {
                '>' => self.beam(x, y - 1, '^'), // reflect
                '<' => self.beam(x, y + 1, 'v'), // reflect
                'v' => self.beam(x - 1, y, '<'), // reflect
                '^' => self.beam(x + 1, y, '>'), // reflect
                _ => panic!(),
            },

            Some('-') => match d {
                '>' => self.beam(x + 1, y, d), // traverse
                '<' => self.beam(x - 1, y, d), // traverse
                'v' | '^' => {
                    self.beam(x - 1, y, '<'); // split the beam
                    self.beam(x + 1, y, '>'); // split the beam
                }
                _ => panic!(),
            },

            Some('|') => match d {
                'v' => self.beam(x, y + 1, d), // traverse
                '^' => self.beam(x, y - 1, d), // traverse
                '>' | '<' => {
                    self.beam(x, y - 1, '^'); // split the beam
                    self.beam(x, y + 1, 'v'); // split the beam
                }
                _ => panic!(),
            },

            _ => panic!(),
        }
    }

    /// Get the number of energized cells.
    fn energized(&self) -> usize {
        self.beams
            .iter()
            .map(|row| row.iter().filter(|&&e| e != 0).count())
            .sum()
    }

    /// Turn on the beam light and get the number of energized cells.
    fn calc(&mut self, x: usize, y: usize, d: char) -> usize {
        self.reset_beams();
        self.beam(x, y, d);
        self.energized()
    }

    /// Solve part one.
    fn part1(&mut self) -> usize {
        self.calc(1, 1, '>')
    }

    /// Solve part two.
    fn part2(&mut self) -> usize {
        let mut m = 0;

        m = m.max(
            (1..=self.sx)
                .map(|x| self.calc(x, 1, 'v').max(self.calc(x, self.sy, '^')))
                .max()
                .unwrap(),
        );

        m = m.max(
            (1..=self.sy)
                .map(|y| self.calc(1, y, '>').max(self.calc(self.sx, y, '<')))
                .max()
                .unwrap(),
        );

        m
    }
}

fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(args.path.as_str());
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test01() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test.txt");
        assert_eq!(puzzle.part1(), 46);
    }

    #[test]
    fn test02() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test.txt");
        assert_eq!(puzzle.part2(), 51);
    }
}
